### PR TITLE
Update flow past a backward facing step example

### PR DESCRIPTION
**Pull request by lpsaavedra**
### Description

The post-processing scripts were not working for this example and the example was long and hard to follow. This PR refactors the example by limiting it to two Reynolds numbers: 100 and 1000.  To summarize:
- All the parameters are revisited to ensure correct convergence for both cases.  
-  The folders and parameter files are renamed so that it is clear for which Reynolds they are.
- The results are limited to: visualizing the velocity profiles (with streamlines to identify eddy after step and zoom in plots, all in one figure), reporting the reattachment point for the last step and its relative error in comparison to reference data, and comparing the obtained velocity distribution at the outlet to the analytical solution.
-  The post-processing scripts are fixed so that they work for both Reynolds number by just changing the appropriate flag when running the script. Moreover, the number of vtu files is no longer hardcoded and can be found automatically by the script.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [X] Lethe documentation is up to date
- [X] Copyright headers are present and up to date
- [X] The branch is rebased onto master
- [X] Code is indented with indent-all and .prm files (examples and tests) with prm-indent
- [X] Links are added to parent .rst files
- [X] The example is following the [standard format](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#general-rules-and-format)

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 1]
```diff
@@ -34,7 +34,7 @@ In this example, a bidimensional flow goes past a backward-facing step. The flow
...
+The backward-facing step problem is a classical computational fluid dynamics problem. The fact that it features a non-trivial solution while maintaining simple geometry and boundary conditions makes this problem a good candidate for validation purposes as well as to test the robustness of a given CFD method. First, the basic parameters used to solve the backward-facing step problem will be exposed. A solution to two Reynolds numbers (:math:`\mathrm{Re} = 100` to :math:`\mathrm{Re} =1000`) will then be presented and compared to experimental and numerical data. A mesh adaptation and numerical error analysis will be carried on.
```
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:37]:
```suggestion
The backward-facing step is a classical computational fluid dynamics problem. The fact that it features a non-trivial solution while maintaining simple geometry and boundary conditions makes this problem a good candidate for validation purposes as well as to test the robustness of a given CFD method. First, the basic parameters of the backward-facing step problem will be shown. A solution to two Reynolds numbers (:math:`\mathrm{Re} = 100`  to :math:`\mathrm{Re} =1000`) will then be presented and compared to experimental and numerical data. A mesh adaptation and numerical error analysis will be carried out.
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 2]
```diff
@@ -46,7 +46,7 @@ The following subsections show the different parameters used in the simulations.
 Simulation Control
 ~~~~~~~~~~~~~~~~~~
 
-For :math:`Re < 700`, the solution is stable enough to be computed in steady state by setting ``method`` to ``steady``: 
+For :math:`\mathrm{Re} = 100`, the solution is stable enough to be computed in steady state by setting ``method`` to ``steady``:
```
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:49]:
```suggestion
At :math:`\mathrm{Re} = 100`, the solution is stable enough to be computed as steady state. We do so by setting ``method`` to ``steady``: 
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 3]
```diff
@@ -55,31 +55,26 @@ For :math:`Re < 700`, the solution is stable enough to be computed in steady sta
...
-      set subdivision       = 1
-      set output boundaries = false
     end
 	
 A mesh refinement analysis can be done with ``set number mesh adapt = 10``. By starting from a very coarse mesh and by dynamically refining the mesh at least 10 times, asymptotic convergence can be clearly observed.
 
-However, for :math:`Re \geq 700`, convergence can be quite difficult to obtain while doing a steady state simulation. In fact, as the Reynolds number increases, the problem becomes progressively stiffer to a point where the ``steady`` solver ultimately fails. With that in mind, the case can be solved as a transient problem until the steady state solution is obtained. This can be achieved with the ``method = steady_bdf`` parameter.
+However, for :math:`\mathrm{Re} = 1000`, convergence can be quite difficult to obtain while doing a steady state simulation. In fact, as the Reynolds number increases, the problem becomes progressively stiffer to a point where the ``steady`` solver ultimately fails. With that in mind, the case can be solved as a transient problem until the steady state solution is obtained. This can be achieved with the ``method = steady_bdf`` parameter.
```
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:62]:
```suggestion
However, at :math:`\mathrm{Re} = 1000`, convergence can be quite difficult to obtain while doing a steady state simulation. As the Reynolds number increases, the problem becomes progressively stiffer to a point where the ``steady`` solver ultimately fails. With that in mind, the case can be solved as a transient problem until the steady state solution is obtained. This can be achieved with the ``method = steady_bdf`` parameter.
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 4]
```diff
@@ -55,31 +55,26 @@ For :math:`Re < 700`, the solution is stable enough to be computed in steady sta
...
-      set subdivision                  = 1
-      set output boundaries            = false
     end
   
-``stop tolerance``, ``time step``, ``adapt``, ``max cfl`` and ``adaptive time step scaling`` are parameters that control the pseudo-steady simulation. In this case, choosing ``stop tolerance = 1e-6`` ensures that the simulation reaches steady state while keeping the number of time iterations to a minimum. Moreover, one can notice a very high value for the ``max cfl``; however, since it is used with ``adaptative time step scaling`` (and since *Lethe* is an implicit solver), even a very high value of the CFL does not compromise the results.
+``stop tolerance``, ``time step``, ``adapt``, ``max cfl`` and ``adaptive time step scaling`` are parameters that control the pseudo-steady simulation. In this case, choosing ``stop tolerance = 1e-5`` ensures that the simulation reaches steady state while keeping the number of time iterations to a minimum. Moreover, one can notice a very high value for the ``max cfl``; however, since it is used with ``adaptative time step scaling`` (and since *Lethe* is an implicit solver), even a very high value of the CFL does not compromise the results.
```
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:77]:
```suggestion
``stop tolerance``, ``time step``, ``adapt``, ``max cfl`` and ``adaptive time step scaling`` are parameters that control the pseudo-steady state simulation. In this case, choosing ``stop tolerance = 1e-5`` ensures that the simulation reaches steady state while keeping the number of time iterations to a minimum. Moreover, one can notice a very high value for the ``max cfl``; however, since it is used with ``adaptative time step scaling`` (and since *Lethe* is an implicit solver), even a very high value of the CFL does not compromise the results.
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 5]
```diff
@@ -135,19 +129,7 @@ In this example, the mesh adaptation algorithm is based on the Kelly error estim
...
-Initial coarse mesh:
-
-.. image:: image/0th-mesh.png
-
-After four refinement steps:
-
-.. image:: image/4th-mesh.png
-
-After eight refinement steps:
+For higher Reynolds number with adjoint time stepping, ``frequency = 5`` can be added to the above parameters in order to obtain a reasonable number of elements throughout the simulation. In this particular case, the mesh would be refined at every fifth time iteration. As an example, the mesh after eight refinement steps for :math:`\mathrm{Re} = 100` looks as follows:
```
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:132]:
```suggestion
For higher Reynolds numbers with adjoint time stepping, ``frequency = 5`` can be added to the above parameters to obtain a reasonable number of elements throughout the simulation. In this particular case, the mesh would be refined at every fifth time iteration. As an example, the mesh after eight refinement steps for :math:`\mathrm{Re} = 100` looks as follows:
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 6]
```diff
@@ -255,115 +240,77 @@ For :math:`Re \geq 700`, however, it is often necessary to set ``ilu precondtion
...
+It is possible to notice that there seems to be a lot of diffusion past the step. This phenomenon is coherent with what is known of the Navier-Stokes equations: the diffusivity term is inversely proportional to the Reynolds number. Most importantly, a small eddy adjacent to the step is clearly observable. It is also visually noticeable that :math:`2.7 \leq x_r \leq 2.9` (:math:`17.7 \leq x \leq 17.9`). With the Python module `PyVista <https://docs.pyvista.org/>`_, raw simulation data can be extracted (from the .vtu files) and this data can be used to compute :math:`x_r` numerically using the following equation:
```
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:263]:
```suggestion
It is possible to notice a lot of diffusion past the step. This phenomenon is coherent with what is known of the Navier-Stokes equations: the diffusivity term is inversely proportional to the Reynolds number. Most importantly, a small eddy adjacent to the step is clearly observable. It is also visually noticeable that :math:`2.7 \leq x_r \leq 2.9` (:math:`17.7 \leq x \leq 17.9`). With the Python module `PyVista <https://docs.pyvista.org/>`_, raw simulation data can be extracted (from the .vtu files) and this data can be used to compute :math:`x_r` numerically using the following equation:
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 7]
```diff
@@ -255,115 +240,77 @@ For :math:`Re \geq 700`, however, it is often necessary to set ``ilu precondtion
...
-+===============+================+================+
-| Order 1       | :math:`10.219` | :math:`1.26` % |
-+---------------+----------------+----------------+
-| Order 2       | :math:`10.250` | :math:`0.96` % |       
-+---------------+----------------+----------------+
-| Erturk (2008) | :math:`10.349` | N/A            |
-+---------------+----------------+----------------+
+On the contrary of what we saw in the :math:`\mathrm{Re} = 100` case, it is clearly noticeable that there is much less diffusion within the flow. This is once more coherent with the theory. The same eddy as mentioned in the previous section is still present, but grows as the Reynolds number is increased. Furthermore, a second principal eddy can be seen adjacent to the top wall in the range :math:`x \in [25,37]`. This "oscillating flow" characteristic is expected of a higher Reynolds flow such as this one. Finally, the :math:`x_r` variable is evaluated visually at :math:`x_r \simeq 12.5` (:math:`x \simeq 27.5`). The same Python code as before can be used by setting the Reynolds flag to ``-Re 1000``; we obtain :math:`x_r = 12.602` as a numerical result with a relative error of :math:`3.9\%`.
```
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:285]:
```suggestion
On the contrary of what we saw in the :math:`\mathrm{Re} = 100` case, it is clear that there is much less diffusion within the flow. This is once more coherent with the theory. The same eddy as mentioned in the previous section is still present but grows as the Reynolds number is increased. Furthermore, a second principal eddy can be seen adjacent to the top wall in the range :math:`x \in [25,37]`. This "oscillating flow" characteristic is expected of a higher Reynolds flow such as this one. Finally, the :math:`x_r` variable is evaluated visually at :math:`x_r \simeq 12.5` (:math:`x \simeq 27.5`). The same Python code as before can be used by setting the Reynolds flag to ``-Re 1000``; we obtain :math:`x_r = 12.602` as a numerical result with a relative error of :math:`3.9\%`.
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/image/Reynolds1000_profile.png
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/image/Reynolds1000_profile.png:1]: Lovely!

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/image/Reynolds100_profile.png
Comment (voferreira) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/image/Reynolds100_profile.png:1]: Lovely! (1)

**Review by voferreira — APPROVED** [review]
Nice images! All clear. I added small suggestions but nothing fancy.

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 8]
```diff
@@ -255,126 +240,83 @@ For :math:`Re \geq 700`, however, it is often necessary to set ``ilu precondtion
...
 
 .. code-block:: text
   :class: copy-button
 
-  mpirun -np j lethe-fluid 2D-backward-facing-step-steady.prm
-	
-where ``j`` is the number of CPUs used for the computations.
+  mpirun -np j lethe-fluid Reynolds100_steady.prm
 
+For the case where :math:`\textrm{Re}=1000`, replace the name of the parameter file by ``Reynolds1000_steadybdf.prm``.
```
Comment (OGaboriault) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:250]: same thing here, with or without the bfd

**Review by OGaboriault — APPROVED** [review]

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 9]
```diff
@@ -20,10 +20,10 @@ All files mentioned below are located in the example's folder (``examples/incomp
 
 - Geometry file: ``backward-facing-step.geo``
 - Mesh file: ``backward-facing-step.msh``
-- Parameter file for the base case (:math:`Re < 700`): ``Reynolds100-600.prm``
-- Parameter file for the higher-Reynolds case (:math:`Re \geq 700`): ``Reynolds700-1000.prm``
-- Postprocessing Python script for computing the reattachment length: ``bisection.py``
-- Postprocessing Python script for computing velocity distributions at inlet and outlet: ``velocity_distribution.py``
+- Parameter file for the base case (:math:`\mathrm{Re} = 100`): ``Reynolds100_steady.prm``
+- Parameter file for the higher-Reynolds case (:math:`\mathrm{Re} = 1000`): ``Reynolds1000_steadybdf.prm``
```
[the remark this answers is not in the document]
Comment (blaisb) [reply doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:24]: no I think this is fine right now no?

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst [hunk 10]
```diff
@@ -55,31 +55,26 @@ For :math:`Re < 700`, the solution is stable enough to be computed in steady sta
...
-      set output boundaries            = false
     end
   
-``stop tolerance``, ``time step``, ``adapt``, ``max cfl`` and ``adaptive time step scaling`` are parameters that control the pseudo-steady simulation. In this case, choosing ``stop tolerance = 1e-6`` ensures that the simulation reaches steady state while keeping the number of time iterations to a minimum. Moreover, one can notice a very high value for the ``max cfl``; however, since it is used with ``adaptative time step scaling`` (and since *Lethe* is an implicit solver), even a very high value of the CFL does not compromise the results.
+``stop tolerance``, ``time step``, ``adapt``, ``max cfl`` and ``adaptive time step scaling`` are parameters that control the pseudo-steady state simulation. In this case, choosing ``stop tolerance = 1e-5`` ensures that the simulation reaches steady state while keeping the number of time iterations to a minimum. Moreover, one can notice a very high value for the ``max cfl``; however, since it is used with ``adaptative time step scaling`` (and since *Lethe* is an implicit solver), even a very high value of the CFL does not compromise the results.
```
Comment (blaisb) [inline doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:77]:
```suggestion
``stop tolerance``, ``time step``, ``adapt``, ``max cfl`` and ``adaptive time step scaling`` are parameters that control the pseudo steady-state simulation. In this case, choosing ``stop tolerance = 1e-5`` ensures that the simulation reaches steady state while keeping the number of time iterations to a minimum. Moreover, one can notice a very high value for the ``max cfl``; however, since it is used with ``adaptative time step scaling`` (and since *Lethe* is an implicit solver), even a very high value of the CFL does not compromise the results.
```

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst (hunk 8 above)
Comment (blaisb) [reply doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:250]: nah I think this is ok right now. Because steadybdf is the name of the method

File: doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (lpsaavedra) [reply doc/source/examples/incompressible-flow/2d-backward-facing-step/2d-backward-facing-step.rst:24]: The reason for this is that both cases are solved differently (see simulation control section). The Reynolds 100 case is solved as a steady problem, while the second one uses steady bdf.

**Review by blaisb — APPROVED** [review]
One tiny typo, then ready for merge

**Comment by lpsaavedra** [comment]
Thank you for your review, I have applied all the comments and I will mark this as ready to merge :)
